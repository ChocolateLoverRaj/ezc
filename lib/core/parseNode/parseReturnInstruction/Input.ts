import EnumItemWithData from '../../EnumItemWithData'
import TryParseNode from '../TryParseNode'

interface Input {
  typeParsers: ReadonlyArray<TryParseNode<EnumItemWithData>>
  valueParsers: ReadonlyArray<TryParseNode<EnumItemWithData>>
}

export default Input
