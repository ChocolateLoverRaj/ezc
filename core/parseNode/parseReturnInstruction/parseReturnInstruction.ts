import CoreKeyWord from '../../CoreKeyWord'
import skip from '../../../util/splitAsyncIterator/skip'
import splitAsyncIterator from '../../../util/splitAsyncIterator/splitAsyncIterator'
import CoreNodesWithData from '../CoreNodesWithData'
import CoreNodeType from '../CoreNodeType'
import tryNodeParsers from '../tryNodeParsers/tryNodeParsers'
import TryParseNode from '../TryParseNode'
import Input from './Input'
import checkKeyWord from '../checkKeyWord'

const parseReturnInstruction = (
  { typeParsers, valueParsers }: Input
): TryParseNode<CoreNodesWithData[CoreNodeType.RETURN_INSTRUCTION]> => async stream => {
  const splittedIterator = splitAsyncIterator(stream[Symbol.asyncIterator]())
  const type = {
    enum: CoreNodeType,
    id: CoreNodeType.RETURN_INSTRUCTION
  }

  {
    const error = await checkKeyWord(
      splittedIterator.asyncIterable[Symbol.asyncIterator](),
      type,
      0,
      'Expected return',
      { enum: CoreKeyWord, id: CoreKeyWord.RETURN }
    )
    if (error !== undefined) return error
    skip(splittedIterator, 1)
  }

  const parseTypeResult = await tryNodeParsers(typeParsers)(splittedIterator.asyncIterable)
  if (!parseTypeResult.success) {
    return {
      success: false,
      result: {
        type,
        index: 1,
        message: 'Expected type',
        subAttempts: parseTypeResult.result
      }
    }
  }
  skip(splittedIterator, parseTypeResult.result.length)

  const parseValueResult = await tryNodeParsers(valueParsers)(splittedIterator.asyncIterable)
  if (!parseValueResult.success) {
    return {
      success: false,
      result: {
        type,
        index: 1 + parseTypeResult.result.length,
        message: 'Expected value',
        subAttempts: parseValueResult.result
      }
    }
  }
  skip(splittedIterator, parseValueResult.result.length)

  return {
    success: true,
    result: {
      node: {
        type,
        data: {
          type: parseTypeResult.result.node,
          value: parseValueResult.result.node
        }
      },
      length: 1 + parseTypeResult.result.length + parseValueResult.result.length
    }
  }
}

export default parseReturnInstruction
