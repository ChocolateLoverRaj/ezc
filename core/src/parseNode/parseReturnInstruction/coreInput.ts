import coreTypeParsers from '../coreTypeParsers'
import coreValueParsers from '../coreValueParsers'
import Input from './Input'

const coreInput: Input = {
  typeParsers: coreTypeParsers,
  valueParsers: coreValueParsers
}

export default coreInput
