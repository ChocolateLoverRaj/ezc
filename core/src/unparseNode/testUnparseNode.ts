import TryParseNode from '../parseNode/TryParseNode'
import EnumItemWithData from '../EnumItemWithData'
import UnparsedPartType from './UnparsedPartType'
import unparsedNodeToString from '../unparsedNodeToString/unparsedNodeToString'
import coreInput from '../unparsedNodeToString/coreInput'
import parseAllTokens from '../tryGetToken/parseAllTokens'
import coreTryers from '../tryGetToken/coreTryers'
import arrayToAsyncIterable from 'util/dist/arrayToAsyncIterable/arrayToAsyncIterable.js'

const testUnparseNode = async (
  parseNode: TryParseNode<EnumItemWithData>,
  code: string
): Promise<void> => {
  expect(unparsedNodeToString(coreInput)([{
    type: UnparsedPartType.NODE,
    data: ((await parseNode(
      parseAllTokens(coreTryers)(
        arrayToAsyncIterable([code])) as any))?.result as any).node as EnumItemWithData
  }])).toMatchSnapshot()
}

export default testUnparseNode
