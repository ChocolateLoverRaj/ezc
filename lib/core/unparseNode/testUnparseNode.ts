import TryParseNode from '../parseNode/TryParseNode'
import EnumItemWithData from '../EnumItemWithData'
import UnparsedPartType from './UnparsedPartType'
import unparsedNodeToString from '../unparsedNodeToString/unparsedNodeToString'
import coreInput from '../unparsedNodeToString/coreInput'
import parseAllTokens from '../tryGetToken/parseAllTokens'
import coreTryers from '../tryGetToken/coreTryers'
import arrayToAsyncIterable from '../arrayToAsyncIterable'

const testUnparseNode = async (
  parseNode: TryParseNode<EnumItemWithData>,
  code: string
): Promise<void> => {
  expect(unparsedNodeToString(coreInput)([{
    type: UnparsedPartType.NODE,
    data: (await parseNode(
      parseAllTokens(coreTryers)(
        arrayToAsyncIterable([code])) as any))?.node as EnumItemWithData
  }])).toMatchSnapshot()
}

export default testUnparseNode
