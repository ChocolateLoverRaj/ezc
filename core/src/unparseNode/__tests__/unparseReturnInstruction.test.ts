import parseReturnInstruction from '../../parseNode/parseReturnInstruction/parseReturnInstruction'
import coreParseReturnInstructionInput from '../../parseNode/parseReturnInstruction/coreInput'
import testUnparseNode from '../testUnparseNode'

test('ret i1 0', async () => {
  await testUnparseNode(parseReturnInstruction(coreParseReturnInstructionInput), 'ret i1 0')
})
