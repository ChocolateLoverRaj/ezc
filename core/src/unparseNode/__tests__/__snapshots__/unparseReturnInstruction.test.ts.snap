// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ret i1 0 1`] = `"ret i1 0"`;
